packages/ts/file-router/src/runtime/RouterBuilder.ts

```diff
@@ -0,0 +1,133 @@
+import { transformTreeSync } from '../shared/transformTree.js';
+import type { AgnosticRoute } from '../types.js';
+import { toReactRouter } from './toReactRouter.js';
+
+/**
+ * Deeply merges two lists of routes. If the specific path is already present,
+ * the route is merged, otherwise the new routes are added to the list.
+ *
+ * @param a - The first list of routes.
+ * @param b - The second list of routes.
+ */
+function mergeRoutes(a: RouteObject[], b: RouteObject[]): RouteObject[] {
+  return b.reduce(
+    (result, route) => {
+      const existingRoute = result.find((r) => r.path === route.path);
+      if (existingRoute) {
+        Object.assign(existingRoute, route);
+        existingRoute.children = existingRoute.children
+          ? mergeRoutes(existingRoute.children, route.children ?? [])
+          : route.children;
+      } else {
+        result.push(route);
+      }
+      return result;
+    },
+    [...a],
+  );
+}
+
+/**
+ * A builder for creating a Vaadin-specific router for React with
+ * authentication and server routes support.
+ */
+export class RouterBuilder {
+  #routes: RouteObject[] = [];
+
+  /**
+   * Returns the current list of routes.
+   */
+  get routes(): readonly RouteObject[] {
+    return this.#routes;
+  }
+
+  /**
+   * Adds the given routes to the current list of routes. All the routes are
+   * deeply merged to preserve the path uniqueness.
+   *
+   * @param routes - A list of routes to add to the current list.
+   */
+  withReactRoutes(...routes: RouteObject[]): this {
+    if (!this.#routes.length) {
+      this.#routes = routes;
+    } else {
+      this.#routes = mergeRoutes(this.#routes, routes);
+    }
+
+    return this;
+  }
+
+  /**
+   * Adds the given file routes to the current list of routes. All the routes
+   * are transformed to React RouterObjects and deeply merged to preserve the
+   * path uniqueness.
+   *
+   * @param routes - A list of routes to add to the current list.
+   */
+  withFileRoutes(...routes: readonly AgnosticRoute[]): this {
+    const reactRoutes = routes.map(toReactRouter);
+    this.withReactRoutes(...reactRoutes);
+    return this;
+  }
+
+  /**
+   * Adds the given server route element to each branch of the current list of
+   * routes.
+   *
+   * @param element - The element to add to each branch of the current list of
+   * routes.
+   */
+  withServerRoutes(element: ReactElement): this {
+    const flowRoute: RouteObject = { path: '*', element };
+
+    this.#routes = this.#routes.map((route) =>
+      transformTreeSync<RouteObject, RouteObject>(
+        route,
+        (r) => r.children?.values(),
+        (r, children) =>
+          // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+          ({
+            ...r,
+            children: [...children, flowRoute],
+          }) as RouteObject,
+      ),
+    );
+
+    this.#routes.push(flowRoute);
+
+    return this;
+  }
```

Comment:
This means that `withServerRoutes` must be used at the end of the chain. I'd rather make this method just store the passed element internally and then the recursive append should happen when `get routes` or `build` are used.

Reply:
Good take. But what would we have on `routes` getter then?

Reply:
IMO you keep the original routes in memory and, when someone accesses the `routes` getter or the `build` method, you add the server routes on the fly, if any, and use the result. No need to store, as it is unlikely that two accesses will happen anyway, I guess.

Reply:
Also, is the `routes` getter really useful outside of tests?

Reply:
In that case, there are only two solutions: either we accept that the `withServerRoutes` must be put at the end, or the final value must be cached and invalidated.

Reply:
As we discussed on grooming, the `build` now returns a `RouterConfiguration` instead of a router itself. So, we can get access to the routes and router only when we call `build()`